util/src/overlaydb.rs

```diff
@@ -40,340 +47,167 @@ use kvdb::{Database, DBTransaction};
-			.map(|d| {
-				let r = Rlp::new(&d);
-				(r.at(1).as_val(), r.at(0).as_val())
-			})
+		Ok(ops)
 	}
 
-	/// Put the refs and value of the given key, possibly deleting it from the db.
-	fn put_payload_in_batch(&self, batch: &DBTransaction, key: &H256, payload: (Bytes, u32)) -> bool {
-		if payload.1 > 0 {
-			let mut s = RlpStream::new_list(2);
-			s.append(&payload.1);
-			s.append(&payload.0);
-			batch.put(key, s.as_raw()).expect("Low-level database error. Some issue with your hard disk?");
-			false
-		} else {
-			batch.delete(key).expect("Low-level database error. Some issue with your hard disk?");
-			true
-		}
-	}
+	/// Revert all operations on this object since last commit.
+	pub fn revert(&mut self) { self.overlay.clear() }
 
-	/// Put the refs and value of the given key, possibly deleting it from the db.
-	fn put_payload(&self, key: &H256, payload: (Bytes, u32)) -> bool {
-		if payload.1 > 0 {
-			let mut s = RlpStream::new_list(2);
-			s.append(&payload.1);
-			s.append(&payload.0);
-			self.backing.put(key, s.as_raw()).expect("Low-level database error. Some issue with your hard disk?");
-			false
-		} else {
-			self.backing.delete(key).expect("Low-level database error. Some issue with your hard disk?");
-			true
-		}
+	/// Get the value of the given key.
+	fn payload(&self, key: &H256) -> Option<Bytes> {
+		// TODO [rob] have this and HashDB functions all return Results.
+		self.backing.get(key).expect("Low level database error.").map(|v| v.to_vec())
 	}
 }
 
 impl HashDB for OverlayDB {
 	fn keys(&self) -> HashMap<H256, i32> {
-		let mut ret: HashMap<H256, i32> = HashMap::new();
+		let mut ret = HashMap::new();
+
 		for (key, _) in self.backing.iter() {
-			let h = H256::from_slice(key.deref());
-			let r = self.payload(&h).unwrap().1;
-			ret.insert(h, r as i32);
+			ret.insert(H256::from_slice(&*key), 1);
 		}
 
-		for (key, refs) in self.overlay.keys().into_iter() {
-			let refs = *ret.get(&key).unwrap_or(&0) + refs;
-			ret.insert(key, refs);
+		for (key, refs) in self.overlay.keys() {
+			*ret.entry(key).or_insert(0) += refs;
 		}
+
 		ret
 	}
+
 	fn get(&self, key: &H256) -> Option<&[u8]> {
-		// return ok if positive; if negative, check backing - might be enough references there to make
-		// it positive again.
-		let k = self.overlay.raw(key);
-		match k {
-			Some(&(ref d, rc)) if rc > 0 => Some(d),
-			_ => {
-				let memrc = k.map_or(0, |&(_, rc)| rc);
-				match self.payload(key) {
-					Some(x) => {
-						let (d, rc) = x;
-						if rc as i32 + memrc > 0 {
-							Some(&self.overlay.denote(key, d).0)
-						}
-						else {
-							None
-						}
-					}
-					// Replace above match arm with this once https://github.com/rust-lang/rust/issues/15287 is done.
-					//Some((d, rc)) if rc + memrc > 0 => Some(d),
-					_ => None,
-				}
-			}
+		match self.overlay.raw(key) {
+			Some(&(ref d, rc)) if rc > 0 => { Some(d) }
+			_ => if let Some(x) = self.payload(key) {
```

Comment:
we need to be careful that this "incorrect" behavior is preserved for `DeletionMode::Ignore` since ArchiveDB uses this exact logic. It is only broken for `DeletionMode::Delete`